appserver/web/gf-weld-connector/src/main/java/org/glassfish/weld/connector/WeldSniffer.java

```diff
@@ -16,213 +17,226 @@
+        if (!isWeldArchive && archiveName != null && archiveName.endsWith(EXPANDED_JAR_SUFFIX)) {
+            isWeldArchive = isArchiveCDIEnabled(context, archive, META_INF_BEANS_XML);
+        }
 
-    boolean isWeldArchive = false;
-    // scan for beans.xml in expected locations. If at least one is found without bean-discovery-mode="none", this is
-    // a Weld archive
-    if (isEntryPresent(archive, WeldUtils.WEB_INF)) {
-      isWeldArchive = isArchiveCDIEnabled(context, archive, WeldUtils.WEB_INF_BEANS_XML) ||
-                      isArchiveCDIEnabled(context, archive, WeldUtils.WEB_INF_CLASSES_META_INF_BEANS_XML);
+        // If stand-alone ejb-jar
+        if (!isWeldArchive && isArchiveCDIEnabled(context, archive, META_INF_BEANS_XML)) {
+            isWeldArchive = true;
+        }
 
-      if (!isWeldArchive) {
-        // Check jars under WEB_INF/lib
-        if (isEntryPresent(archive, WeldUtils.WEB_INF_LIB)) {
-          isWeldArchive = scanLibDir(context, archive, WeldUtils.WEB_INF_LIB);
+        if (!isWeldArchive && archiveName != null && archiveName.endsWith(EXPANDED_RAR_SUFFIX)) {
+            isWeldArchive = isArchiveCDIEnabled(context, archive, META_INF_BEANS_XML);
+            if (!isWeldArchive) {
+                // Check jars in root dir of rar
+                isWeldArchive = scanLibDir(context, archive, "");
+            }
         }
-      }
-    }
 
-    // TODO This doesn't seem to match the ReadableArchive for a stand-alone ejb-jar.
-    // It might only be true for an ejb-jar within an .ear.  Revisit when officially
-    // adding support for .ears
-    String archiveName = archive.getName();
-    if (!isWeldArchive && archiveName != null && archiveName.endsWith(WeldUtils.EXPANDED_JAR_SUFFIX)) {
-      isWeldArchive = isArchiveCDIEnabled(context, archive, WeldUtils.META_INF_BEANS_XML);
+        return isWeldArchive;
     }
 
-    // If stand-alone ejb-jar
-    if (!isWeldArchive && isArchiveCDIEnabled(context, archive, WeldUtils.META_INF_BEANS_XML) ) {
-      isWeldArchive = true;
-    }
+    private boolean scanLibDir(DeploymentContext context, ReadableArchive archive, String libLocation) {
+        boolean entryPresent = false;
+
+        if (libLocation != null) {
+            Enumeration<String> entries = archive.entries(libLocation);
+            while (entries.hasMoreElements() && !entryPresent) {
+                String entryName = entries.nextElement();
+                // If a jar in lib dir and not WEB-INF/lib/foo/bar.jar
+                if (entryName.endsWith(JAR_SUFFIX) && entryName.indexOf(SEPARATOR_CHAR, libLocation.length() + 1) == -1) {
+                    try {
+                        ReadableArchive jarInLib = archive.getSubArchive(entryName);
+                        entryPresent = isArchiveCDIEnabled(context, jarInLib, META_INF_BEANS_XML);
+                        jarInLib.close();
+                    } catch (IOException e) {
+                    }
```

Comment:
I would prefer logging the exception instead of ignorance. Even if the log is all we do with that, it is still a good warning/severe message that something unexpected happened.

Reply:
Can try that, although in a next PR I'd also wanted to look at making ReadableArchive autocloseable, so the code will change again.